src/java/org/apache/cassandra/db/SSTableImporter.java

```diff
@@ -97,7 +101,39 @@ synchronized List<String> importNewSSTables(Options options)
                         try
                         {
                             abortIfDraining();
-                            verifySSTableForImport(descriptor, entry.getValue(), options.verifyTokens, options.verifySSTables, options.extendedVerify);
+
+                            if (options.failOnMissingIndex)
+                            {
+                                Index.Group saiIndexGroup = cfs.indexManager.getIndexGroup(StorageAttachedIndexGroup.GROUP_KEY);
+                                if (saiIndexGroup != null)
+                                {
+                                    IndexDescriptor indexDescriptor = IndexDescriptor.create(descriptor,
+                                                                                             cfs.getPartitioner(),
+                                                                                             cfs.metadata().comparator);
+
+                                    String keyspace = cfs.getKeyspaceName();
+                                    String table = cfs.getTableName();
+
+                                    if (!indexDescriptor.isPerSSTableIndexBuildComplete())
+                                        throw new IllegalStateException(String.format("Missing SAI index to import for SSTable %s on %s.%s",
+                                                                                      indexDescriptor.sstableDescriptor.toString(),
+                                                                                      keyspace,
+                                                                                      table));
+
+                                    for (Index index : saiIndexGroup.getIndexes())
+                                    {
+                                        IndexIdentifier indexIdentifier = new IndexIdentifier(keyspace, table, index.getIndexMetadata().name);
```

Comment:
nit: `StorageAttachedIndex` has an `IndexIdentifier` already if you want to cast the `Index`, which should be safe.

Reply:
I rather leave this as it is ... casting this etc ... nah.

Reply:
Sure. `IndexIdentifier` itself could actually be an index implementation-agnostic class. There's nothing ASI-specific forcing it to be in `org.apache.cassandra.index.sai.utils`.